magmaR/R/query.R

```diff
@@ -0,0 +1,99 @@
+#' Analagous to the '/query' fucntion of magma
+#' @inheritParams retrieve
+#' @param queryTerms An array of query predicates, see \url{https://mountetna.github.io/magma.html#query}
+#' @param format Either "list" or "df" (=dataframe), this sets the desired output format
+#' @param ... Additional parameters passed along to the internal `.query()` function.
+#' For troubleshooting purposes only.
+#' Options: request.only (Logical), json.params.only (Logical), verbose (Logical)
+#' @return Either list conversion of the raw '/query' JSON output if \code{format == "list"}, default,
+#' 
+#' OR a dataframe conversion if \code{format = "df"}
+#' @export
+#' @examples
+#' 
+#' ##### FILLER
+#' 
+#' # Unless a working TOKEN is hard-coded, or it is in an interactive session,
+#' #   this code will not work.
+#' 
+#' if (interactive()) {
+#'     # Running like this will ask for input of your janus token.
+#'     retrieveJSON(
+#'         projectName = "ipi",
+#'         modelName = "patient",
+#'         recordNames = "all",
+#'         attributeNames = "all",
+#'         filter = "")
+#' }
+#'
+
+query <- function(
+    projectName,
+    queryTerms = list(),
+    format = c("list", "df"),
+    token = .get_TOKEN(),
+    ...
+) {
+    
+    format <- match.arg(format)
+    
+    raw <- .query(
+        projectName = projectName,
+        queryTerms = queryTerms,
+        token = token,
+        ...)
+    
+    if (format == "list") {
+        raw
+    } else {
+        out <- data.frame(raw$answer)
+        colnames(out) <- raw$format
+        
+        out
+    }
+}
+
+.query <- function(
+    projectName,
+    queryTerms = list(),
+    request.only = FALSE,
+    json.params.only = FALSE,
+    token = .get_TOKEN(),
+    verbose = FALSE
+) {
+    
+    ### Put together the request 
+    
+    jsonParams <- list(
+        project_name = projectName,
+        query = queryTerms)
+
+    requestBody <- jsonlite::toJSON(jsonParams, auto_unbox = TRUE)
+    
+    ### Output here if requested.
+    if (request.only) {
+        return(requestBody)
+    }
+    if (json.params.only) {
+        return(jsonParams)
+    }
+    
+    ### Perform '\query'
+    
+    h <- RCurl::basicTextGatherer()
+    
+    RCurl::curlPerform(
+        url = "https://magma.ucsf.edu/query",
```

Comment:
It might be useful for testing if this value could be configurable in a non-committed / .gitignore config file.

Reply:
Another option: I already have it so that "hidden" options are passable to the `retrieve()` and `query()` functions for troubleshooting, via `...`. They're documented with a 'for troubleshooting only' descriptor, but that could be updated to 'for advanced users', and we could add an input for adjusting the url to the underlying `.retrieve()` and `.query()` functions.  Perhaps `url.base` and it would only modify the "https://magma.ucsf.edu" portion.

Then (useful for DSs) a reinstall of the package wouldn't be required for switching between production vs staging vs dev.

Tokens would still need to match up, so I think this change is safe?  

I'm not sure on the details of how your original suggestion would work, but we could do that too in order to not need to be explicit in every function call in testing. What I'd suggest we target with such an approach would be the default option of the `url.base` input that I'll add in.

Reply:
Oh, I see, like a default value that someone who calls the method can overwrite? Yeah, I think hidden options or default option would work, too, and may make more sense in an R context -- having to uninstall / reinstall seems like a bad workflow. Logically I think most of the work will be against this production Magma URL, but it feels weird to have it hardcoded or to force someone to edit source code if they want to hit a different endpoint.

Reply:
These:
https://github.com/mountetna/monoetna/blob/ff2dbfab9bc1cfae75032183cd127a6776f138ff/magmaR/R/retrieve.R#L115
https://github.com/mountetna/monoetna/blob/ff2dbfab9bc1cfae75032183cd127a6776f138ff/magmaR/R/query.R#L61

Agree that users editing source code is not ideal. I hadn't thought towards how best to add in non-production capability until this discussion so just to be clear, it's not that I was leaning that direction.

Anyway, *should* work now via giving `retrieve/query(..., url.base = "staging_url")`, but would you perhaps be able to check for staging / dev?  I think I'm only set up for dev, but I don't actually have that running right now.

Code for testing would be:
```
# Install current:
remotes::install_github("mountetna/monoetna@magmaR", subdir = "magmaR")

# RESTART your R session (in RStudio, it's in the menu bar Session > Restart R)
library(magmaR)
retrieve(
    projectName = "ipi",
    modelName = "sample",
    recordNames = "all",
    attributeNames = "tissue_type",
    url.base = *staging/dev*,
    token = *appropriate_token*)
```

Reply:
Thanks for the R code, very helpful!

I think making the url configurable with `.base` looks like it works. But this also leads down a rabbit hole I think we should put off until later. The issue will be that staging / dev use self-signed SSL certificates, so you get an error like:

```
Error in function (type, msg, asError = TRUE)  : 
  SSL certificate problem: unable to get local issuer certificate
```

But I think most folks will only need production, and it may not be worth trying to fix at this point.

Reply:
The dev cert uses a custom CA, it is not "self-signed" like staging; we should just convert staging to use the same custom CA.

Reply:
Oh that's a good idea. That should be simple to do -- monoetna has a helper script to generate certs off of that CA.